File: src/main/java/com/example/chatService/service/RoomParticipantServiceImpl.java
package com.example.chatService.service;

import com.example.chatService.dto.AdminChangedResponse;
import com.example.chatService.dto.ParticipantDto;
import com.example.chatService.dto.RoomCountDto;
import com.example.chatService.dto.RoomRole;
import com.example.chatService.entity.ChatRoom;
import com.example.chatService.entity.RoomParticipant;
import com.example.chatService.event.ParticipantForcedExitEvent;
import com.example.chatService.event.RoomParticipantsChangedEvent;
import com.example.chatService.exception.BannedFromRoomException;
import com.example.chatService.repository.ChatRoomV2Repository;
import com.example.chatService.repository.RoomParticipantRepository;
import com.example.chatService.repository.UserRepository;
import jakarta.persistence.OptimisticLockException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.retry.annotation.Retryable;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.PathVariable;

import java.time.Duration;
import java.util.List;
import java.util.Optional;

import static com.example.chatService.dto.RoomRole.*;

@Service
@RequiredArgsConstructor
@Transactional
@Slf4j
public class RoomParticipantServiceImpl implements RoomParticipantService {

    private final RoomParticipantRepository roomParticipantRepository;
    private final UserRepository userRepository;
    private final StringRedisTemplate redis;
    private final ChatRoomV2Repository roomRepository;
    private final ParticipantEventPublisherImpl publisher;
    private final SimpMessagingTemplate messagingTemplate;
    private final ApplicationEventPublisher eventPublisher;

    @Override
    @Retryable(
            value = OptimisticLockException.class,
            maxAttempts = 3
    )
    @Transactional
    public void joinRoom(String roomId, String userId) {

        if (roomParticipantRepository.existsByRoomIdAndUserIdAndIsBannedTrue(roomId, userId)) {
            throw new BannedFromRoomException(roomId);
        }
        // 유저 권한 확인
        RoomRole role = hasPermission(roomId, userId);

        joinAsRole(roomId, userId, role);
    }

    public RoomRole hasPermission(String roomId, String userId) {
        Optional<RoomParticipant> member =
                roomParticipantRepository.findByRoomIdAndUserId(roomId, userId);

        if (member.isEmpty()) return MEMBER;

        return member.get().getRole();
    }
    @Transactional
    public void joinAsRole(String roomId, String userId, RoomRole roomRole) {

//        System.out.println("joinAsRole = " + roomId + " " + userId + " " + roomRole);
        RoomParticipant p = roomParticipantRepository.findByRoomIdAndUserId(roomId, userId)
                .orElseGet(() -> roomParticipantRepository.save(
                        new RoomParticipant(
                                roomId,
                                userId,
                                roomRole
                        )
                ));

//        System.out.println("isActive = " + p.isActive());

        if (p.isActive()) {
            return;
        }

        ChatRoom room = roomRepository.findByIdForUpdate(roomId);
//        System.out.println("[BEFORE] room currentCount = " + room.getCurrentCount());
        room.increaseCount();
//        System.out.println("[AFTER] room currentCount = " + room.getCurrentCount());

        p.activate();
        roomParticipantRepository.save(p);

        eventPublisher.publishEvent(
                new RoomParticipantsChangedEvent(roomId)
        );
    }

    @Override
    @Transactional
    public void leaveRoom(String roomId, String userId) {

        RoomParticipant participant = getParticipant(roomId, userId);
        if (!participant.isActive()) {
            return;
        }

        ChatRoom room = roomRepository.findByIdForUpdate(roomId);
        room.decreaseCount();

        participant.deactivate();
        roomParticipantRepository.save(participant);

        eventPublisher.publishEvent(
                new RoomParticipantsChangedEvent(roomId)
        );
    }

    @Override
    public void reconnect(String roomId, String userId) {
    }

    @Override
    @Transactional
    public void kick(String roomId, String targetUserId, String byUserId) {
        requireAdmin(roomId, byUserId);

        if (byUserId.equals(targetUserId)) {
            throw new IllegalStateException("Cannot kick yourself");
        }

        RoomParticipant target = getParticipant(roomId, targetUserId);

        if (target.getRole() == RoomRole.OWNER) {
            throw new IllegalStateException("OWNER는 강퇴할 수 없습니다.");
        }

        if (!target.isActive()) {
            return;
        }

        target.deactivate();

        roomParticipantRepository.save(target);

        ChatRoom room = roomRepository.findByIdForUpdate(roomId);
        room.decreaseCount();

        eventPublisher.publishEvent(
                new ParticipantForcedExitEvent(
                        roomId,
                        targetUserId,
                        "KICK"
                )
        );

        eventPublisher.publishEvent(
                new RoomParticipantsChangedEvent(roomId)
        );
    }

    @Override
    @Transactional
    public void ban(String roomId, String targetUserId, String byUserId, String reason) {
        requireOwner(roomId, byUserId);

        if (targetUserId.equals(byUserId)) {
            throw new IllegalStateException("Cannot ban yourself");
        }

        RoomParticipant target = roomParticipantRepository
                .findByRoomIdAndUserId(roomId, targetUserId)
                .orElseThrow(() -> new IllegalStateException("Target not in room"));

        if (target.getRole() == RoomRole.OWNER) {
            throw new IllegalStateException("Cannot ban OWNER");
        }

        boolean wasActive = target.isActive();
        target.ban(reason);   // isBanned=true, isActive=false


        if (wasActive) {
            ChatRoom room = roomRepository.findByIdForUpdate(roomId);
            room.decreaseCount();

            roomParticipantRepository.save(target);

            eventPublisher.publishEvent(
                    new ParticipantForcedExitEvent(
                            roomId,
                            targetUserId,
                            reason
                    )
            );
            eventPublisher.publishEvent(
                    new RoomParticipantsChangedEvent(roomId)
            );
        }
    }

    @Override
    @Transactional
    public void changeRole(
            String roomId,
            String targetUserId,
            RoomRole role,
            String byUserId
    ) {
        validateOwner(roomId, byUserId);

        RoomParticipant target = getParticipant(roomId, targetUserId);
        target.changeRole(role);

        roomParticipantRepository.save(target);
    }

    @Override
    @Transactional
    public AdminChangedResponse toggleAdmin(
            String roomId,
            String requesterId,
            String targetUserId
    ) {
        RoomParticipant requester = roomParticipantRepository
                .findByRoomIdAndUserId(roomId, requesterId)
                .orElseThrow();

        if (requester.getRole() != RoomRole.OWNER)
            throw new AccessDeniedException("OWNER만 가능");

        RoomParticipant target = roomParticipantRepository
                .findByRoomIdAndUserId(roomId, targetUserId)
                .orElseThrow();

        if (target.getRole() == RoomRole.OWNER)
            throw new IllegalStateException("OWNER는 변경 불가");

        target.setRole(
                target.getRole() == ADMIN ? MEMBER : ADMIN
        );

        roomParticipantRepository.save(target);

        return new AdminChangedResponse(
                targetUserId,
                target.getRole().name()
        );
    }


    @Override
    @Transactional
    public void transferOwnership(
            String roomId,
            String newOwnerId,
            String byUserId
    ) {
        requireOwner(roomId, byUserId);

        ChatRoom room = roomRepository.findByIdForUpdate(roomId);

        if (!room.getOwnerUserId().equals(byUserId)) {
            new SecurityException("Owner only");
        }

        if (byUserId.equals(newOwnerId)) return;
        room.setOwnerUserId(newOwnerId);
        roomRepository.save(room);

        publisher.broadcastOwnerChanged(roomId, newOwnerId);
    }

    private void requireOwner(String roomId, String userId) {
        ChatRoom room = roomRepository.findById(roomId).orElseThrow();

        if (!room.getOwnerUserId().equals(userId)) {
            throw new SecurityException("OWNER only");
        }
    }

    private void requireAdmin(String roomId, String userId) {
        RoomParticipant p = roomParticipantRepository
                .findByRoomIdAndUserId(roomId, userId)
                .orElseThrow(() -> new AccessDeniedException("Not in room"));

        if (p.getRole() == RoomRole.MEMBER) {
            throw new AccessDeniedException("ADMIN only");
        }
    }
    @Override
    @Transactional(readOnly = true)
    public List<RoomParticipant> getActiveParticipants(String roomId) {
        return roomParticipantRepository.findAllByRoomIdAndIsActiveTrue(roomId);
    }

    @Override
    public List<ParticipantDto> getParticipants(@PathVariable String roomId) {
        return getActiveParticipants(roomId)
                .stream()
                .map(p -> new ParticipantDto(
                        p.getUserId(),
                        loadUsername(p.getUserId()),
                        p.getRole()
                ))
                .toList();
    }

    private RoomParticipant getParticipant(String roomId, String userId) {
        return roomParticipantRepository.findByRoomIdAndUserId(roomId, userId)
                .orElseThrow(() -> new IllegalStateException(
                        "Participant not found. roomId=" + roomId + ", userId=" + userId
                ));
    }

    private void validateOwner(String roomId, String userId) {
        RoomParticipant p = getParticipant(roomId, userId);
        if (p.getRole() != OWNER) {
            throw new SecurityException("Owner only");
        }
    }
    @Override
    @Transactional
    public int getCurrentCount(String roomId) {
        return roomRepository.findById(roomId)
                .orElseThrow()
                .getCurrentCount();
    }

    private String loadUsername(String userId) {

        if (userId == null) return "UNKNOWN";

        String key = "user:" + userId + ":username";
        String cached = redis.opsForValue().get(key);

        if (cached != null) return cached;

        String fromDb = userRepository.findUsernameValueById(userId);

        if (fromDb == null) {
            log.warn("Username not found for userId={}", userId);
            return "UNKNOWN";
        }
        redis.opsForValue().set(key, fromDb, Duration.ofHours(1));
        return fromDb;
    }

    private ParticipantDto toDto(RoomParticipant p) {
        return new ParticipantDto(
                p.getUserId(),
                loadUsername(p.getUserId()),
                p.getRole()
        );
    }

    @Override
    @Transactional
    public void broadcast(String roomId) {
        int current = getCurrentCount(roomId);

        ChatRoom room = roomRepository.findById(roomId)
                .orElseThrow(() -> new RuntimeException("broadcast"));

        messagingTemplate.convertAndSend(
                "/topic/room-users/" + roomId,
                "UPDATED"
        );

        RoomCountDto dto = new RoomCountDto(
                current,
                room.getMaxParticipants()
        );
        messagingTemplate.convertAndSend(
                "/topic/rooms/" + roomId + "/count",
                dto
        );
    }
    public boolean isParticipant(String roomId, String userId) {
        return roomParticipantRepository.existsByRoomIdAndUserId(roomId, userId);
    }
}

File: src/main/java/com/example/chatService/service/RoomParticipantServiceImpl.kt
package com.example.chatService.service

import com.example.chatService.dto.AdminChangedResponse
import com.example.chatService.dto.ParticipantDto
import com.example.chatService.dto.RoomCountDto
import com.example.chatService.dto.RoomRole
import com.example.chatService.dto.RoomRole.ADMIN
import com.example.chatService.dto.RoomRole.MEMBER
import com.example.chatService.dto.RoomRole.OWNER
import com.example.chatService.entity.ChatRoom
import com.example.chatService.entity.RoomParticipant
import com.example.chatService.event.ParticipantForcedExitEvent
import com.example.chatService.event.RoomParticipantsChangedEvent
import com.example.chatService.exception.BannedFromRoomException
import com.example.chatService.repository.ChatRoomV2Repository
import com.example.chatService.repository.RoomParticipantRepository
import com.example.chatService.repository.UserRepository
import jakarta.persistence.OptimisticLockException
import org.slf4j.LoggerFactory
import org.springframework.context.ApplicationEventPublisher
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.retry.annotation.Retryable
import org.springframework.security.access.AccessDeniedException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Duration

@Service
@Transactional
class RoomParticipantServiceImpl(

    private val roomParticipantRepository: RoomParticipantRepository,
    private val userRepository: UserRepository,
    private val redis: StringRedisTemplate,
    private val roomRepository: ChatRoomV2Repository,
    private val publisher: ParticipantEventPublisherImpl,
    private val messagingTemplate: SimpMessagingTemplate,
    private val eventPublisher: ApplicationEventPublisher,

) : RoomParticipantService {
    @Retryable(value = [OptimisticLockException::class], maxAttempts = 3)
    @Transactional
    override fun joinRoom(roomId: String, userId: String) {
        if (roomParticipantRepository.existsByRoomIdAndUserIdAndIsBannedTrue(roomId, userId)) {
            throw BannedFromRoomException(roomId)
        }

        val role = hasPermission(roomId, userId)
        joinAsRole(roomId, userId, role)
    }

    fun hasPermission(roomId: String, userId: String): RoomRole {
        val member = roomParticipantRepository.findByRoomIdAndUserId(roomId, userId)
        if (member.isEmpty) {
            return MEMBER
        }
        return member.get().role ?: MEMBER
    }

    @Transactional
    fun joinAsRole(roomId: String, userId: String, roomRole: RoomRole) {
        val participant = roomParticipantRepository.findByRoomIdAndUserId(roomId, userId)
            .orElseGet {
                roomParticipantRepository.save(
                    RoomParticipant(roomId = roomId, userId = userId, role = roomRole),
                )
            }

        if (participant.isActive) {
            return
        }

        val room = roomRepository.findByIdForUpdate(roomId)
        room.increaseCount()
        participant.activate()
        roomParticipantRepository.save(participant)

        eventPublisher.publishEvent(RoomParticipantsChangedEvent(roomId))
    }

    @Transactional
    override fun leaveRoom(roomId: String, userId: String) {
        val participant = getParticipant(roomId, userId)
        if (!participant.isActive) {
            return
        }

        val room = roomRepository.findByIdForUpdate(roomId)
        room.decreaseCount()
        participant.deactivate()
        roomParticipantRepository.save(participant)
        eventPublisher.publishEvent(RoomParticipantsChangedEvent(roomId))
    }

    override fun reconnect(roomId: String, userId: String) {
    }

    @Transactional
    override fun kick(roomId: String, targetUserId: String, byUserId: String) {
        requireAdmin(roomId, byUserId)
        if (byUserId == targetUserId) {
            throw IllegalStateException("Cannot kick yourself")
        }

        val target = getParticipant(roomId, targetUserId)
        if (target.role == RoomRole.OWNER) {
            throw IllegalStateException("OWNER는 강퇴할 수 없습니다.")
        }
        if (!target.isActive) {
            return
        }

        target.deactivate()
        roomParticipantRepository.save(target)

        val room = roomRepository.findByIdForUpdate(roomId)
        room.decreaseCount()

        eventPublisher.publishEvent(ParticipantForcedExitEvent(roomId, targetUserId, "KICK"))
        eventPublisher.publishEvent(RoomParticipantsChangedEvent(roomId))
    }

    @Transactional
    override fun ban(roomId: String, targetUserId: String, byUserId: String, reason: String) {
        requireOwner(roomId, byUserId)
        if (targetUserId == byUserId) {
            throw IllegalStateException("Cannot ban yourself")
        }

        val target = roomParticipantRepository.findByRoomIdAndUserId(roomId, targetUserId)
            .orElseThrow { IllegalStateException("Target not in room") }

        if (target.role == RoomRole.OWNER) {
            throw IllegalStateException("Cannot ban OWNER")
        }

        val wasActive = target.isActive
        target.ban(reason)

        if (wasActive) {
            val room = roomRepository.findByIdForUpdate(roomId)
            room.decreaseCount()
            roomParticipantRepository.save(target)

            eventPublisher.publishEvent(ParticipantForcedExitEvent(roomId, targetUserId, reason))
            eventPublisher.publishEvent(RoomParticipantsChangedEvent(roomId))
        }
    }

    @Transactional
    override fun changeRole(roomId: String, targetUserId: String, role: RoomRole, byUserId: String) {
        validateOwner(roomId, byUserId)
        val target = getParticipant(roomId, targetUserId)
        target.changeRole(role)
        roomParticipantRepository.save(target)
    }

    @Transactional
    override fun toggleAdmin(roomId: String, requesterId: String, targetUserId: String): AdminChangedResponse {
        val requester = roomParticipantRepository.findByRoomIdAndUserId(roomId, requesterId).orElseThrow()
        if (requester.role != RoomRole.OWNER) {
            throw AccessDeniedException("OWNER만 가능")
        }

        val target = roomParticipantRepository.findByRoomIdAndUserId(roomId, targetUserId).orElseThrow()
        if (target.role == RoomRole.OWNER) {
            throw IllegalStateException("OWNER는 변경 불가")
        }

        target.role = if (target.role == ADMIN) MEMBER else ADMIN
        roomParticipantRepository.save(target)

        return AdminChangedResponse(targetUserId, target.role?.name ?: MEMBER.name)
    }

    @Transactional
    override fun transferOwnership(roomId: String, newOwnerId: String, byUserId: String) {
        requireOwner(roomId, byUserId)
        val room = roomRepository.findByIdForUpdate(roomId)

        if (room.ownerUserId != byUserId) {
            throw SecurityException("Owner only")
        }
        if (byUserId == newOwnerId) {
            return
        }

        room.ownerUserId = newOwnerId
        roomRepository.save(room)
        publisher.broadcastOwnerChanged(roomId, newOwnerId)
    }

    private fun requireOwner(roomId: String, userId: String) {
        val room = roomRepository.findById(roomId).orElseThrow()
        if (room.ownerUserId != userId) {
            throw SecurityException("OWNER only")
        }
    }

    private fun requireAdmin(roomId: String, userId: String) {
        val participant = roomParticipantRepository.findByRoomIdAndUserId(roomId, userId)
            .orElseThrow { AccessDeniedException("Not in room") }

        if (participant.role == RoomRole.MEMBER) {
            throw AccessDeniedException("ADMIN only")
        }
    }

    @Transactional(readOnly = true)
    override fun getActiveParticipants(roomId: String): List<RoomParticipant> =
        roomParticipantRepository.findAllByRoomIdAndIsActiveTrue(roomId)

    override fun getParticipants(roomId: String): List<ParticipantDto> =
        getActiveParticipants(roomId).map {
            ParticipantDto(
                it.userId,
                loadUsername(it.userId),
                it.role ?: MEMBER,
            )
        }

    private fun getParticipant(roomId: String, userId: String): RoomParticipant =
        roomParticipantRepository.findByRoomIdAndUserId(roomId, userId)
            .orElseThrow {
                IllegalStateException("Participant not found. roomId=$roomId, userId=$userId")
            }

    private fun validateOwner(roomId: String, userId: String) {
        val participant = getParticipant(roomId, userId)
        if (participant.role != OWNER) {
            throw SecurityException("Owner only")
        }
    }

    @Transactional
    override fun getCurrentCount(roomId: String): Int =
        roomRepository.findById(roomId).orElseThrow().currentCount

    private fun loadUsername(userId: String?): String {
        if (userId == null) {
            return "UNKNOWN"
        }

        val key = "user:$userId:username"
        val cached = redis.opsForValue().get(key)
        if (cached != null) {
            return cached
        }

        val fromDb = userRepository.findUsernameValueById(userId)
        if (fromDb == null) {
            log.warn("Username not found for userId={}", userId)
            return "UNKNOWN"
        }

        redis.opsForValue().set(key, fromDb, Duration.ofHours(1))
        return fromDb
    }

    @Transactional
    override fun broadcast(roomId: String) {
        val current = getCurrentCount(roomId)
        val room: ChatRoom = roomRepository.findById(roomId)
            .orElseThrow { RuntimeException("broadcast") }

        messagingTemplate.convertAndSend("/topic/room-users/$roomId", "UPDATED")
        messagingTemplate.convertAndSend(
            "/topic/rooms/$roomId/count",
            RoomCountDto(current, room.maxParticipants),
        )
    }

    override fun isParticipant(roomId: String, userId: String): Boolean =
        roomParticipantRepository.existsByRoomIdAndUserId(roomId, userId)

    companion object {
        private val log = LoggerFactory.getLogger(RoomParticipantServiceImpl::class.java)
    }
}
